### PR TITLE
Rewrite of Google Images plug-in

### PR DESCRIPTION
**Pull request by GrosPoulet**
This new version of plug-in catches "on the fly" data sent by Google servers when user scrolls down images results.
Note: this method could be applied to other plug-ins, removing the need for API calls.

### CODE REVIEW

File: plugins/google.js [hunk 1]
```diff
@@ -26,35 +27,132 @@ hoverZoomPlugins.push({
...
-        // Extract direct links to images using document.scripts data (no api call(s) needed)
-        // Mostly inspired from chocolatebot's userscript : Google DWIMages 
+        // Hook Google 'Open' XMLHttpRequests to catch data & metadata associated with pictures displayed
+        // These requests are issued by client to Google servers in order to obtain new data when user scrolls down
+        // Hooked data is stored in scripts in DOM with class='hoverZoomHookedData'
+        if ($('script.hoverZoomHook').length == 0) { // Inject hook script in document if not already there
+            var hookScript = document.createElement('script');
+            hookScript.type = 'text/javascript';
+            hookScript.text = `if (typeof oldXHROpen !== 'function') { // Hook only once!
+                console.log('Hooking XHROpen');
```
Comment (extesy) [inline plugins/google.js:37]: Please don't pollute console with log messages. You should use `cLog` instead (defined in hoverzoom.js).

File: plugins/google.js [hunk 2]
```diff
@@ -26,35 +27,132 @@ hoverZoomPlugins.push({
...
-        function parseScripts() {
-            scripts = Array.from(document.scripts);
-            callbacks = scripts.filter(script => /^AF_initDataCallback\b/.test(script.text));
-            dataFromScripts = callbacks.pop().text;
-        };
 
-        parseScripts();
+        // Extract data in initDataCallback
+        function extractInitData(tbnid) {
+            console.log('extractInitData');
```
Comment (extesy) [inline plugins/google.js:86]: cLog

File: plugins/google.js [hunk 3]
```diff
@@ -26,35 +27,132 @@ hoverZoomPlugins.push({
...
+                let lastquoteIndex = initData.indexOf('"', firstquoteIndex + 1);
+                url = initData.substring(firstquoteIndex + 1, lastquoteIndex);
+            }
+            url = cleanUrl(url);
+            return url;
+        }
+
+        // Extract hooked data stored in <script> elements appended to DOM
+        function extractHookedData(tbnid) {
+            console.log('extractHookedData');
```
Comment (extesy) [inline plugins/google.js:107]: cLog

File: plugins/google.js [hunk 4]
```diff
@@ -26,35 +27,132 @@ hoverZoomPlugins.push({
...
+                            (document.head || document.documentElement).appendChild(fakeA);
+                            (document.head || document.documentElement).removeChild(fakeA);
+                        } catch {}
+                    });
+                    // Proceed with original function
+                    return oldXHROpen.apply(this, arguments);
+                }
+            }`;
+            hookScript.classList.add('hoverZoomHook');
+            (document.head || document.documentElement).appendChild(hookScript);
```
Comment (extesy) [inline plugins/google.js:62]: I don't understand why do you need to dynamically create a script. All scripts are already executed by the browser, such as for example this `plugins/google.js` file.

File: plugins/google.js [hunk 5]
```diff
@@ -26,35 +27,132 @@ hoverZoomPlugins.push({
...
+                    this.addEventListener('load', function() {
+                        try {
+                            //console.log('load: ' + this.responseText);
+                            // store response as plain text in a <script> element for later usage by plug-in
+                            var hookedDataScript = document.createElement('script');
+                            hookedDataScript.type = 'text';
+                            hookedDataScript.text = this.responseText;
+                            hookedDataScript.classList.add('hoverZoomHookedData'); // add specific class for easier retrieval
+                            (document.head || document.documentElement).appendChild(hookedDataScript);
+                            // Add & remove empty <a> element to/from DOM to trigger HoverZoom,
```
Comment (extesy) [inline plugins/google.js:50]: Instead of doing that hack, why not call appropriate hoverzoom function directly?

File: plugins/google.js [hunk 6]
```diff
@@ -26,35 +27,132 @@ hoverZoomPlugins.push({
...
+            hookScript.type = 'text/javascript';
+            hookScript.text = `if (typeof oldXHROpen !== 'function') { // Hook only once!
+                console.log('Hooking XHROpen');
+                oldXHROpen = window.XMLHttpRequest.prototype.open;
+                window.XMLHttpRequest.prototype.open = function(method, url, async, user, password) {
+                    // catch responses
+                    this.addEventListener('load', function() {
+                        try {
+                            //console.log('load: ' + this.responseText);
+                            // store response as plain text in a <script> element for later usage by plug-in
```
Comment (extesy) [inline plugins/google.js:44]: Why do you need to store it as script instead of using a `data-` attribute?

File: plugins/google.js [hunk 7]
```diff
@@ -26,35 +27,132 @@ hoverZoomPlugins.push({
...
+                initData = goodScripts.pop().text;
+            }
+            let tbnidq = '"' + tbnid + '",';
+            if (initData.indexOf(tbnidq) != -1) {
+                let firstquoteIndex = initData.indexOf('[', initData.indexOf('[', initData.indexOf(tbnidq)) + 1) + 1;
+                let lastquoteIndex = initData.indexOf('"', firstquoteIndex + 1);
+                url = initData.substring(firstquoteIndex + 1, lastquoteIndex);
+            }
+            url = cleanUrl(url);
+            return url;
```
Comment (extesy) [inline plugins/google.js:102]: These two lines can be combined.

File: plugins/google.js (hunk 1 above)
Comment (GrosPoulet) [reply plugins/google.js:37]: ok, done

File: plugins/google.js (hunk 2 above)
Comment (GrosPoulet) [reply plugins/google.js:86]: ok, done

File: plugins/google.js (hunk 3 above)
Comment (GrosPoulet) [reply plugins/google.js:107]: ok, done

File: plugins/google.js (hunk 7 above)
Comment (GrosPoulet) [reply plugins/google.js:102]: ok, done

File: plugins/google.js (hunk 4 above)
Comment (GrosPoulet) [reply plugins/google.js:62]: It's a script with type = 'text', so it'll not be executed by browser.
It's just a place to store chunks of data returned by google when user scrolls down results.

File: plugins/google.js (hunk 5 above)
Comment (GrosPoulet) [reply plugins/google.js:50]: Yes, it's a hack to trigger the function: prepareImgLinksAsync(dontResetDelay)
I can change it, but this function would have to become public ?

File: plugins/google.js (hunk 6 above)
Comment (GrosPoulet) [reply plugins/google.js:44]: Some chunks of data to store are quite bigs (around 30/50 Ko), so i don't know if datasets are appropriate... 
What's your opinion?

Comment (extesy) [reply plugins/google.js:44]: You can also use a local (preferably) or global variable, similar to existing `hoverZoomPlugins` and `hoverZoom`. That would be even better.

File: plugins/google.js (hunk 5 above)
Comment (extesy) [reply plugins/google.js:50]: I think calling prepareImgLinksAsync would be better. It's ok to make it public.

File: plugins/google.js (hunk 4 above)
Comment (extesy) [reply plugins/google.js:62]: Javascript variables can hold state until the page is closed, you don't need to use DOM to store data.

Comment (GrosPoulet) [reply plugins/google.js:62]: I modified this part: XHR responses are now stored in sessionStorage. So no more modifications to DOM needed.

File: plugins/google.js (hunk 5 above)
Comment (GrosPoulet) [reply plugins/google.js:50]: Unfortunatly, the code that intercepts XHR responses is executed <b>in the page context</b>, not in the extension context. So i did not manage to call prepareImgLinksAsync from here, maybe you have a solution?

File: plugins/google.js (hunk 6 above)
Comment (GrosPoulet) [reply plugins/google.js:44]: I now use sessionStorage, no need for dataset updates or variables.

**Comment by GrosPoulet** [comment]
No, i was waiting for your advices on lines 47-50

File: plugins/google.js (hunk 5 above)
Comment (extesy) [reply plugins/google.js:50]: Hm, ok, let's keep it as-is for now and maybe we'll come up with a less hacky approach.

**Comment by extesy** [comment]
Ah, sorry, I don't have any good ideas right now so we can ignore that comment.

**Pull request merged**